test/http_test.js

```diff
@@ -259,21 +259,48 @@ describe("HTTP class", () => {
     });
 
     describe("Retry-After header handling", () => {
-      beforeEach(() => {
-        // Make Date#getTime always returning 1000000, for predictability
-        sandbox.stub(Date.prototype, "getTime").returns(1000 * 1000);
-        sandbox.stub(events, "emit");
-      });
+      describe("Event", () => {
+        beforeEach(() => {
+          // Make Date#getTime always returning 1000000, for predictability
+          sandbox.stub(Date.prototype, "getTime").returns(1000 * 1000);
+          sandbox.stub(events, "emit");
+        });
 
-      it("should emit a retry-after event when Retry-After is set", () => {
-        sandbox.stub(global, "fetch").returns(
-          fakeServerResponse(200, {}, {"Retry-After": "1000"}));
+        it("should emit a retry-after event when Retry-After is set", () => {
+          sandbox.stub(global, "fetch").returns(
+            fakeServerResponse(200, {}, {"Retry-After": "1000"}));
 
-        return http.request("/").then(_ => {
-          expect(events.emit.lastCall.args[0]).eql("retry-after");
-          expect(events.emit.lastCall.args[1]).eql(2000000);
+          return http.request("/", {retry: 0}).then(_ => {
+            expect(events.emit.lastCall.args[0]).eql("retry-after");
+            expect(events.emit.lastCall.args[1]).eql(2000000);
+          });
         });
       });
+
+      describe("Retry loop", () => {
+        let fetch;
+
+        beforeEach(() => {
+          fetch = sandbox.stub(global, "fetch");
+        })
+
+        it("should retry the request once by default", () => {
+          const success = {success: true};
+          fetch.onCall(0).returns(fakeServerResponse(503, {}, {"Retry-After": "1"}));
+          fetch.onCall(1).returns(fakeServerResponse(200, success));
+          return http.request("/")
+            .then(res => res.json)
+            .should.eventually.become(success);
+        }).timeout(1100);
+
+        it("should error when retries are exhausted", () => {
+          fetch.onCall(0).returns(fakeServerResponse(503, {}, {"Retry-After": "1"}));
+          fetch.onCall(1).returns(fakeServerResponse(503, {}, {"Retry-After": "1"}));
+          fetch.onCall(2).returns(fakeServerResponse(503, {}, {"Retry-After": "1"}));
+          return http.request("/", {retry: 2})
+            .should.eventually.be.rejectedWith(Error, /Error: HTTP 503/);
+        }).timeout(2100);
```

Comment:
This is very ugly indeed, it is just to control mocha timeout for this specific test.

I didn't succeed to use fake timers, because I had to `clock.tick()` once the errored response was received. I tried to tick from a `events.on('retry-after')` callback but it had no effect.

IMO this is really bad because it really slows down the tests suite execution!!